### PR TITLE
Report non-GROUP BY columns in select's Subquery

### PR DESCRIPTION
**Pull request by findepi**
Find and report column references within a Subquery in a select
expression that not part of the GROUP BY and therefore should not be
used. Such queries now cause a SemanticException, previously they would
fail with a RuntimeException when building a plan.

Outstanding work:

* allow un-GROUP-ed-BY columns within a select's Subquery if they appear
    in an aggregation, e.g.

        WITH t(a,b) as (VALUES (1,2))
        SELECT (SELECT count(a)) FROM t GROUP BY b;

        WITH t(a,b) as (VALUES (1,2))
        SELECT (SELECT 1 FROM t WHERE t.a = count(u.a)) FROM t u GROUP BY b;

* in case of GROUP BY <expression>, allow the expression in a select's
    Subquery, e.g.

        WITH t(a,b) as (VALUES (1,2))
        SELECT (SELECT b+1) FROM t GROUP BY b+1;

### CODE REVIEW

**Comment by findepi** [comment]
@martint, this fixes #7030 but leaves several problems unaddressed. I'm going to create separate issues for the outstanding work items described above. Is that OK?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/Select.java [hunk 1]
```diff
@@ -64,7 +64,7 @@ public boolean isDistinct()
     @Override
     public List<Node> getChildren()
     {
-        return null;
+        return ImmutableList.copyOf(selectItems);
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/Select.java]: No need for a copy here, since it's already an immutable list. (the copy is a no-op, so technically it's fine, but we don't do this on other parts of the code so it's confusing)

Comment (findepi) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/Select.java]: Thanks for explanation. I would feel uncomfortable simply returning `selectItems` (what if someone changes field type...) unless the method return type is narrowed to `ImmutableList`. I hope that's OK.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java [hunk 2]
```diff
@@ -167,9 +173,34 @@ protected Boolean visitAtTimeZone(AtTimeZone node, Void context)
         @Override
         protected Boolean visitSubqueryExpression(SubqueryExpression node, Void context)
         {
+            /*
+             * Walk the subquery and check usages of our fields (the fields defined in our scope) that are not part of GROUP BY.
+             */
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: I would remove this comment

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java [hunk 3]
```diff
@@ -167,9 +173,34 @@ protected Boolean visitAtTimeZone(AtTimeZone node, Void context)
...
 
+        private void analyzeSubqueryReference(Expression referenceExpression)
+        {
+            ResolvedField resolvedField = analysis.getScope(referenceExpression)
+                    .tryResolveField(referenceExpression)
+                    // Unresolvable fields should be found and reported earlier.
+                    .orElseThrow(() -> new IllegalStateException(format("Cannot resolve %s", referenceExpression)));
+
+            if (resolvedField.getScope() != scope) {
+                // Not our scope, not our responsibility
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: and left this one as it is funny ;)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java [hunk 4]
```diff
@@ -167,9 +173,34 @@ protected Boolean visitAtTimeZone(AtTimeZone node, Void context)
...
+                    .forEach(this::analyzeSubqueryReference);
+
             return true;
         }
 
+        private void analyzeSubqueryReference(Expression referenceExpression)
+        {
+            ResolvedField resolvedField = analysis.getScope(referenceExpression)
+                    .tryResolveField(referenceExpression)
+                    // Unresolvable fields should be found and reported earlier.
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: can you please point where? Not for me, but for other readers of this code.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java [hunk 5]
```diff
@@ -167,9 +173,34 @@ protected Boolean visitAtTimeZone(AtTimeZone node, Void context)
...
+                    .orElseThrow(() -> new IllegalStateException(format("Cannot resolve %s", referenceExpression)));
+
+            if (resolvedField.getScope() != scope) {
+                // Not our scope, not our responsibility
+                return;
+            }
+            if (!fieldIndexes.contains(resolvedField.getFieldIndex())) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: I don't like this name `fieldIndexes`. It has no semantic information. What would you say to name it `groupingFields`?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 6]
```diff
@@ -1727,7 +1727,15 @@ private void verifyAggregations(
             Expression expression,
             Set<Expression> columnReferences)
     {
-        AggregationAnalyzer analyzer = new AggregationAnalyzer(groupByExpressions, metadata, scope, columnReferences, analysis.getParameters(), analysis.isDescribe());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: extract this as separate commit

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/Select.java (hunk 1 above)
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/Select.java]: It should belong to separate commit

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/Select.java [hunk 7]
```diff
@@ -62,9 +62,9 @@ public boolean isDistinct()
     }
 
     @Override
-    public List<Node> getChildren()
+    public ImmutableList<SelectItem> getChildren()
```
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/Select.java]: e e... Please return `List`. That is the convention.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java (hunk 4 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: `ExpressionAnalyzer` does that, but I'm reluctant in putting this information here. My intention is that: _if_ after some code changes the exception is being raised in the next line _then_ either we need proper check-n-report here _or_ think more about flows.

In short, the exception servers only as an explanation why I don't do proper reporting here.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java (hunk 3 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: Thanks ;)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 6 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: This is not reformatting. I added `analysis` here and didn't want to have this long line even longer.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java (hunk 5 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: Done

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/Select.java (hunk 7 above)
Comment (findepi) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/Select.java]: Done

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/Node.java [hunk 8]
```diff
@@ -40,7 +40,7 @@ protected Node(Optional<NodeLocation> location)
         return location;
     }
 
-    public abstract List<Node> getChildren();
+    public abstract List<? extends Node> getChildren();
```
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/Node.java]: you can rebase this PR to remove commits which are already merged

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java [hunk 9]
```diff
@@ -342,10 +342,10 @@ protected Boolean visitFunctionCall(FunctionCall node, Void context)
                 }
             }
             else if (node.getFilter().isPresent()) {
-                    throw new SemanticException(MUST_BE_AGGREGATION_FUNCTION,
-                            node,
-                            "Filter is only valid for aggregation functions",
-                            node);
+                throw new SemanticException(MUST_BE_AGGREGATION_FUNCTION,
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: convention is to put each arg in separate line

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java [hunk 10]
```diff
@@ -167,9 +173,31 @@ protected Boolean visitAtTimeZone(AtTimeZone node, Void context)
         @Override
         protected Boolean visitSubqueryExpression(SubqueryExpression node, Void context)
         {
+            AstUtils.preOrder(node)
+                    .filter(Predicates.isInstanceOfAny(DereferenceExpression.class, Identifier.class))
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: static import for `Predicates`?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java (hunk 9 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: I didn't change this. The indentation was wrong.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java (hunk 10 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: Okay

**Comment by findepi** [comment]
Rebased on current `master`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: fair point :)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java (hunk 9 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: and comment is also about indentation

**Comment by sopel39** [comment]
I wonder if the refactor: https://github.com/prestodb/presto/pull/7255 doesn't simplify this PR. Please wait for my review

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java [hunk 11]
```diff
@@ -167,9 +173,31 @@ protected Boolean visitAtTimeZone(AtTimeZone node, Void context)
...
+                    .filter(isInstanceOfAny(DereferenceExpression.class, Identifier.class))
+                    .map(Expression.class::cast)
+                    .forEach(this::analyzeSubqueryReference);
+
             return true;
         }
 
+        private void analyzeSubqueryReference(Expression referenceExpression)
+        {
+            ResolvedField resolvedField = analysis.getScope(referenceExpression)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: why do you assume every `DereferenceExpression` or `Identifier` has a scope? It doesn't seem to be the case

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: Oh I see, it works because of `GetScopeVisitor`, but it might return output scope of wrapping node (e.g: output scope of `QuerySpecificaction` for an select expression), which you shouldn't resolve select symbol against.

For instance, `select (a + 1) as a from ...`. This code would resolve expression `a` from `a+1` against output scope, so it will basically resolve back to `a+1`. Even if this works in the case of this check, this is fishy at least. 

I think analysis should store `ResolvedField` explicitly for every `DefererenceExpression`, `Identity` and `FieldReference`. It should be filled by `ExpressionAnalyzer`. Then it would be easy to check if reference points to grouping fields from correct scope. I have similar problem in my ongoing work for implicit coercions for ORDER BY expressions. I am going to solve the issue as described.

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: The case that might not work:
`SELECT (SELECT a AS a) FROM (VALUES (1,2)) t(a,b) GROUP BY b`

Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: Indeed, thanks!

**Comment by findepi** [comment]
The approach presented here is too simple to solve the problem. I'm still working on the fix but will create a new PR when have something new to show.

**Comment by findepi** [comment]
Superseded by #7354
